pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/instance/JavaInstanceRunnable.java

```diff
@@ -317,46 +322,77 @@ private void setupStateStore() throws Exception {
         }
     }
 
-    private void processResult(Record srcRecord,
-                               CompletableFuture<JavaExecutionResult> result) throws Exception {
-        result.whenComplete((result1, throwable) -> {
-            if (throwable != null || result1.getUserException() != null) {
-                Throwable t = throwable != null ? throwable : result1.getUserException();
-                log.warn("Encountered exception when processing message {}",
-                        srcRecord, t);
-                stats.incrUserExceptions(t);
-                srcRecord.fail();
-            } else {
-                if (result1.getResult() != null) {
-                    sendOutputMessage(srcRecord, result1.getResult());
+    private void processResult(@SuppressWarnings("rawtypes") Record srcRecord,
+                               JavaExecutionResult result) throws SinkException {
+    	
+    	if (result.getUserException() != null) {
+            stats.incrUserExceptions(result.getUserException());
+            srcRecord.fail();
+            return;
+    	}
+    	
+    	if (result.isAsync()) {
+          result.getFuture().whenComplete((result1, throwable) -> {
+              if (throwable != null) {
+                  Throwable t = throwable;
+                  log.warn("Encountered exception when processing message {}", srcRecord, t);
+                  stats.incrUserExceptions(t);
+                  srcRecord.fail();
+              } else {
+                if (result1 != null) {
+                  try {
+                    sendOutputMessage(srcRecord, result1);
+                  } catch (SinkException e) {
+                    log.warn("Encountered exception when publishing result {}", srcRecord, e);
+                  }
                 } else {
-                    if (instanceConfig.getFunctionDetails().getAutoAck()) {
-                        // the function doesn't produce any result or the user doesn't want the result.
-                        srcRecord.ack();
-                    }
+                  if (instanceConfig.getFunctionDetails().getAutoAck()) {
+                    // the function doesn't produce any result or the user doesn't want the result.
+                    srcRecord.ack();
+                  }
                 }
                 // increment total successfully processed
                 stats.incrTotalProcessedSuccessfully();
+              }
+          }); 
+         } else {
```

Comment:
In this code path we still need to check if there was a user exception thrown in the function and fail the record is there was.

Reply:
Whether sync or async the error handling should be consistent:

https://github.com/apache/pulsar/pull/10370/files#diff-4b164878665f89af44f9ace0a3df8a27f4f2fb700fd9981d663fb3a7e4317e7aR338